sys/dev/isp/isp.c

```diff
@@ -4662,15 +4858,468 @@ isp_parse_flt_2xxx(ispsoftc_t *isp, uint8_t *flt_data)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	struct isp_image_status pri_image_status, sec_image_status;
+	bool valid_pri_image = false, valid_sec_image = false;
+	bool active_pri_image = false, active_sec_image = false;
+
+	if (!fcp->flt_region_img_status_pri) {
+		isp_prt(isp, ISP_LOGWARN,
+		    "Primary image not addressed");
+		goto check_sec_image;
+	}
+
+	if (isp_read_flash_data(isp, (uint32_t *) &pri_image_status,
+	    fcp->flt_region_img_status_pri, sizeof(pri_image_status) >> 2) !=
+	    ISP_SUCCESS)
+		goto check_sec_image;
+
+	isp_print_image(isp, "Primary image", &pri_image_status);
+
+	if (isp_check_image_status_signature(&pri_image_status)) {
+		isp_prt(isp, ISP_LOGERR,
+		    "Primary image signature (0x%x) not valid",
+		    le32toh(pri_image_status.signature));
+		goto check_sec_image;
+	}
+
+	if (isp_image_status_checksum(&pri_image_status)) {
+		isp_prt(isp, ISP_LOGERR,
+		    "Primary image checksum failed");
+		goto check_sec_image;
+	}
+
+	valid_pri_image = true;
+
+	if (pri_image_status.image_status_mask & 1) {
+		isp_prt(isp, ISP_LOGCONFIG,
+		    "Primary image is active");
+		active_pri_image = true;
+	}
+
+check_sec_image:
+	if (!fcp->flt_region_img_status_sec) {
+		isp_prt(isp, ISP_LOGWARN,
+		    "Secondary image not addressed");
+		return;
+	}
+
+	if (isp_read_flash_data(isp, (uint32_t *) &sec_image_status,
+	    fcp->flt_region_img_status_sec, sizeof(sec_image_status) >> 2) !=
+	    ISP_SUCCESS)
+		return;
+
+	isp_print_image(isp, "Secondary image", &sec_image_status);
+
+	if (isp_check_image_status_signature(&sec_image_status)) {
+		isp_prt(isp, ISP_LOGERR,
+		    "Secondary image signature (0x%x) not valid",
+		    le32toh(sec_image_status.signature));
+	}
+
+	if (isp_image_status_checksum(&sec_image_status)) {
+		isp_prt(isp, ISP_LOGERR,
+		    "Secondary image checksum failed");
+		goto check_valid_image;
+	}
+
+	valid_sec_image = true;
+
+	if (sec_image_status.image_status_mask & 1) {
+		isp_prt(isp, ISP_LOGCONFIG,
+		    "Secondary image is active");
+		active_sec_image = true;
+	}
+
+check_valid_image:
+	if (valid_pri_image && active_pri_image)
+		active_regions->global = ISP27XX_PRIMARY_IMAGE;
+
+	if (valid_sec_image && active_sec_image) {
+		if (!active_regions->global ||
+		    isp_compare_image_generation(isp,
+			&pri_image_status, &sec_image_status) < 0) {
+			active_regions->global = ISP27XX_SECONDARY_IMAGE;
+		}
+	}
+
+	isp_prt(isp, ISP_LOGDEBUG0, "active image %s (%u)",
+	    active_regions->global == ISP27XX_DEFAULT_IMAGE ?
+		"default (boot/fw)" :
+	    active_regions->global == ISP27XX_PRIMARY_IMAGE ?
+		"primary" :
+	    active_regions->global == ISP27XX_SECONDARY_IMAGE ?
+		"secondary" : "invalid",
+	    active_regions->global);
+}
+
+static bool isp_risc_firmware_invalid(ispsoftc_t *isp, uint32_t *dword)
+{
+	return
+	    !(dword[4] | dword[5] | dword[6] | dword[7]) ||
+	    !(~dword[4] | ~dword[5] | ~dword[6] | ~dword[7]);
```

Comment:
Also misses parentheses.  And since it is not boolean ' == 0' would be better.

Reply:
style(9) explicitly tells "Do not use ! for tests unless it is a boolean".  `!`, `!!`, etc are considered Linux'isms.

Reply:
The bit manipulations look weird, but a bit better.  You do not need one set of parentheses in the first line.

Reply:
fixed